pkg/queue/ring_buffer.go

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) 2020 The Jaeger Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package queue
+
+type RingBufferQueue struct {
```

Comment:
This also require the consumer to take care of concurrency. Would also be good to have a documentation somewhere showing how the types relate to each other. Like: 

> ConcurrentMemoryBoundedQueue wraps MemoryBoundedQueue, which wraps RingBufferQueue

Reply:
or we can make these package private and export just the concurrent version.